src/structs.rs

```diff
@@ -208,4 +211,96 @@ impl Values {
+            GameMode::Mania =>
+                ((self.hit_geki + self.hit_300) as f64 * 6. + self.hit_katu as f64 * 4. + self.hit_100 as f64 * 2. + self.hit_50 as f64)
+                / ((self.hit_geki + self.hit_300 + self.hit_katu + self.hit_100 + self.hit_50 + self.hit_miss) as f64 * 6.)
+        }
+    }
+
+    pub fn get_current_grade(&self) -> String {
+        let total_hits = self.passed_objects as f64;
+        let base_grade = match self.gameplay_gamemode() {
+            GameMode::Osu => {
+                let ratio300 = self.hit_300 as f64 / total_hits;
+                let ratio50 = self.hit_50 as f64 / total_hits;
+                if self.accuracy == 1. {
+                    "SS"
+                } else if ratio300 > 0.9 && self.hit_miss == 0 && ratio50 <= 0.1 {
+                    "S"
+                } else if ratio300 > 0.8 && self.hit_miss == 0 || ratio300 > 0.9 {
+                    "A"
+                } else if ratio300 > 0.7 && self.hit_miss == 0 || ratio300 > 0.8 {
+                    "B"
+                } else if ratio300 > 0.6 {
+                    "C"
+                } else {
+                    "D"
+                }
+            },
+            GameMode::Taiko => {
+                let ratio300 = self.hit_300 as f64 / total_hits;
+                if self.accuracy == 1. {
+                    "SS"
+                } else if ratio300 > 0.9 && self.hit_miss == 0 {
+                    "S"
+                } else if ratio300 > 0.8 && self.hit_miss == 0 || ratio300 > 0.9 {
+                    "A"
+                } else if ratio300 > 0.7 && self.hit_miss == 0 || ratio300 > 0.8 {
+                    "B"
+                } else if ratio300 > 0.6 {
+                    "C"
+                } else {
+                    "D"
+                }
+            },
+            GameMode::Catch => {
+                if self.accuracy == 1. {
+                    "SS"
+                } else if self.accuracy > 0.98 {
+                    "S"
+                } else if self.accuracy > 0.94 {
+                    "A"
+                } else if self.accuracy > 0.90 {
+                    "B"
+                } else if self.accuracy > 0.85 {
+                    "C"
+                } else {
+                    "D"
+                }
+            },
+            GameMode::Mania => {
+                if self.accuracy == 1. {
+                    "SS"
+                } else if self.accuracy > 0.95 {
+                    "S"
+                } else if self.accuracy > 0.9 {
+                    "A"
+                } else if self.accuracy > 0.8 {
+                    "B"
+                } else if self.accuracy > 0.7 {
+                    "C"
+                } else {
+                    "D"
+                }
+            }
+        };
+        // Hidden | Flashlight | Fade In
+        let silver = if self.mods & (8 | 1024 | 1048576) > 0 { "H" } else { "" };
+        format!("{}{}", base_grade, silver)
```

Comment:
Can't talk about the performance difference between reading from memory and calculating manually but if this does go through, it might be better to avoid the `String` allocation by returning a `&'static str` somewhat like this instead
```rs
match (base_grade, mods & (8 | 1024 | 1048576)) {
    ("SS", conj) if conj > 0 => "SSH",
    ("S", conj) if conj > 0 => "SH",
    _ => base_grade,
}
```

Reply:
*imagine missing SS/S check smh*

Thanks!